Link against static libnettle

There's no libnettle udeb.

--- a/configure.ac
+++ b/configure.ac
@@ -1725,7 +1725,7 @@ fi
 if test "x$with_sha1" = xlibnettle; then
 	AC_DEFINE([HAVE_SHA1_IN_LIBNETTLE], [1],
 	          [Use libnettle SHA1 functions])
-	SHA1_LIBS=-lnettle
+	SHA1_LIBS=-l:libnettle.a
 fi
 AC_CHECK_LIB([gcrypt], [gcry_md_open], [HAVE_LIBGCRYPT=yes])
 if test "x$with_sha1" = x && test "x$HAVE_LIBGCRYPT" = xyes; then
