Index: xorg-server-1.1.0/hw/xfree86/os-support/bus/linuxPci.c
===================================================================
--- xorg-server-1.1.0.orig/hw/xfree86/os-support/bus/linuxPci.c	2006-04-28 22:28:56.000000000 +0000
+++ xorg-server-1.1.0/hw/xfree86/os-support/bus/linuxPci.c	2006-08-06 14:02:36.000000000 +0000
@@ -740,8 +740,10 @@
 	write(fd, "1", 2);
 	lseek(fd, 0, SEEK_SET);
 
+    len = min(Len, st.st_size);
+
         /* copy the ROM until we hit Len, EOF or read error */
-        for (i = 0; i < Len && read(fd, Buf, 1) > 0; Buf++, i++)
+        for (; len && (size = read(fd, Buf, len)) > 0 ; Buf+=size, len-=size)
             ;
 
 	write(fd, "0", 2);
