Index: xorg-server/hw/xfree86/common/xf86DPMS.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86DPMS.c	2008-01-06 15:59:07.000000000 -0500
+++ xorg-server/hw/xfree86/common/xf86DPMS.c	2008-01-06 16:09:03.000000000 -0500
@@ -83,7 +83,6 @@
 	    && !DPMSDisabledSwitch)
 	    DPMSEnabled = TRUE;
 	xf86MarkOptionUsed(DPMSOpt);
-	xf86DrvMsg(pScreen->myNum, X_CONFIG, "DPMS enabled\n");
     } else if (DPMSEnabledSwitch) {
 	if (!DPMSDisabledSwitch)
 	    DPMSEnabled = TRUE;
@@ -92,6 +91,8 @@
     else {
 	pDPMS->Enabled = defaultDPMSEnabled;
     }
+    if (pDPMS->Enabled)
+	xf86DrvMsg(pScreen->myNum, X_CONFIG, "DPMS enabled\n");
     pDPMS->CloseScreen = pScreen->CloseScreen;
     pScreen->CloseScreen = DPMSClose;
     DPMSCount++;
