Index: xorg-server/man/Xserver.man
===================================================================
--- xorg-server.orig/man/Xserver.man
+++ xorg-server/man/Xserver.man
@@ -186,10 +186,10 @@ Prohibit creating indirect GLX contexts.
 since it lacks support for many modern OpenGL features and extensions;
 it's slower than direct contexts; and it opens a large attack surface for
 protocol parsing errors.
+This is the default unless +iglx is specified.
 .TP 8
 .B +iglx
 Allow creating indirect GLX contexts.
-This is the default unless \-iglx is specified.
 .TP 8
 .B \-maxbigreqsize \fIsize\fP
 sets the maximum big request to
