Index: xorg-server-1.0.2/hw/xfree86/common/xf86Config.c
===================================================================
--- xorg-server-1.0.2.orig/hw/xfree86/common/xf86Config.c	2006-04-15 17:01:15.000000000 +0800
+++ xorg-server-1.0.2/hw/xfree86/common/xf86Config.c	2006-04-15 17:02:43.000000000 +0800
@@ -616,9 +616,11 @@
     if (fileconf->file_fontpath) {
       char *f = xf86ValidateFontPath(fileconf->file_fontpath);
       pathFrom = X_CONFIG;
-      if (*f)
-        defaultFontPath = f;
-      else {
+      if (*f) {
+        char *fp = xnfalloc(strlen(defaultFontPath) + strlen(f) + 2);
+        sprintf(fp, "%s,%s", f, defaultFontPath);
+        defaultFontPath = fp;
+      } else {
 	xf86Msg(X_WARNING,
 	    "FontPath is completely invalid.  Using compiled-in default.\n");
         fontPath = NULL;
