Some drivers need XFree86Server defined.
Not submitted upstream.

Index: xorg-server/include/xorg-server.h.in
===================================================================
--- xorg-server.orig/include/xorg-server.h.in	2007-07-24 20:57:34.000000000 +0200
+++ xorg-server/include/xorg-server.h.in	2007-07-24 21:12:36.000000000 +0200
@@ -238,4 +238,7 @@
 /* Loadable XFree86 server awesomeness */
 #undef XFree86LOADER
 
+/* Some drivers need that */
+#undef XFree86Server
+
 #endif /* _XORG_SERVER_H_ */
