--- a/hw/dmx/glxProxy/glxext.h	2007-04-30 16:25:07.000000000 +0100
+++ b/hw/dmx/glxProxy/glxext.h	2007-04-30 16:25:43.000000000 +0100
@@ -67,7 +67,7 @@
 
 extern void __glXNoSuchRenderOpcode(GLbyte*);
 extern int __glXNoSuchSingleOpcode(__GLXclientState*, GLbyte*);
-extern void __glXErrorCallBack(__GLinterface *gc, GLenum code);
+extern void __glXErrorCallBack(GLenum code);
 extern void __glXClearErrorOccured(void);
 extern GLboolean __glXErrorOccured(void);
 extern void __glXResetLargeCommandStatus(__GLXclientState*);
