Index: xorg-server/configure.ac
===================================================================
--- xorg-server.orig/configure.ac	2006-11-26 03:00:42.000000000 +0100
+++ xorg-server/configure.ac	2006-11-26 03:05:03.000000000 +0100
@@ -383,6 +383,9 @@
 AC_ARG_WITH(rgb-path,         AS_HELP_STRING([--with-rgb-path=PATH], [Path to RGB database (default: ${datadir}/X11/rgb)]),
 				[ RGBPATH="$withval" ],
 				[ RGBPATH="${datadir}/X11/rgb" ])
+AC_ARG_WITH(serverconfig-path, AS_HELP_STRING([--with-serverconfig-path=PATH], [Path to server config (default: ${libdir}/xserver)]),
+               [ SERVERCONFIG="$withval" ],
+               [ SERVERCONFIG="${libdir}/xserver" ])
 AC_ARG_WITH(dri-driver-path,  AS_HELP_STRING([--with-dri-driver-path=PATH], [Path to DRI drivers (default: ${libdir}/dri)]),
 				[ DRI_DRIVER_PATH="$withval" ],
 				[ DRI_DRIVER_PATH="${libdir}/dri" ])
@@ -821,6 +824,7 @@
 
 AC_DEFINE_DIR(COMPILEDDEFAULTFONTPATH, FONTPATH, [Default font path])
 AC_DEFINE_DIR(RGB_DB, RGBPATH, [Default RGB path])
+AC_DEFINE_DIR(SERVERCONFIGdir, SERVERCONFIG, [Server config path])
 AC_DEFINE_DIR(BASE_FONT_PATH, FONTDIR, [Default base font path])
 AC_DEFINE_DIR(DRI_DRIVER_PATH, DRI_DRIVER_PATH, [Default DRI driver path])
 AC_DEFINE_UNQUOTED(XVENDORNAME, ["$VENDOR_STRING"], [Vendor name])
Index: xorg-server/Xext/Makefile.am
===================================================================
--- xorg-server.orig/Xext/Makefile.am	2006-11-13 19:59:22.000000000 +0100
+++ xorg-server/Xext/Makefile.am	2006-11-26 03:05:59.000000000 +0100
@@ -34,7 +34,6 @@
 	xcmisc.c
 
 # Extra configuration files ship with some extensions
-SERVERCONFIGdir = $(libdir)/xserver
 SERVERCONFIG_DATA =
 
 # Optional sources included if extension enabled by configure.ac rules
