Index: xorg-server/Xi/stubs.c
===================================================================
--- xorg-server.orig/Xi/stubs.c	2010-03-30 13:57:21.000000000 +0300
+++ xorg-server/Xi/stubs.c	2010-03-30 13:57:21.000000000 +0300
@@ -227,7 +227,13 @@
  *
  */
 int
-NewInputDeviceRequest(InputOption *options, InputAttributes *attrs,
+NewInputDeviceRequest(InputOption *options, DeviceIntPtr *pdev)
+{
+    return BadValue;
+}
+
+int
+NewInputDeviceRequest18(InputOption *options, InputAttributes *attrs,
                       DeviceIntPtr *pdev)
 {
     return BadValue;
Index: xorg-server/config/dbus.c
===================================================================
--- xorg-server.orig/config/dbus.c	2010-03-30 13:57:21.000000000 +0300
+++ xorg-server/config/dbus.c	2010-03-30 13:57:21.000000000 +0300
@@ -147,7 +147,7 @@
         dbus_message_iter_next(&iter);
     }
 
-    ret = NewInputDeviceRequest(options, NULL, &dev);
+    ret = NewInputDeviceRequest18(options, NULL, &dev);
     if (ret != Success) {
         DebugF("[config/dbus] NewInputDeviceRequest failed\n");
         goto unwind;
Index: xorg-server/config/hal.c
===================================================================
--- xorg-server.orig/config/hal.c	2010-03-30 13:57:21.000000000 +0300
+++ xorg-server/config/hal.c	2010-03-30 13:57:21.000000000 +0300
@@ -359,7 +359,7 @@
 
     /* this isn't an error, but how else do you output something that the user can see? */
     LogMessage(X_INFO, "config/hal: Adding input device %s\n", name);
-    if ((rc = NewInputDeviceRequest(options, &attrs, &dev)) != Success) {
+    if ((rc = NewInputDeviceRequest18(options, &attrs, &dev)) != Success) {
         LogMessage(X_ERROR, "config/hal: NewInputDeviceRequest failed (%d)\n", rc);
         dev = NULL;
         goto unwind;
Index: xorg-server/config/udev.c
===================================================================
--- xorg-server.orig/config/udev.c	2010-03-30 13:57:21.000000000 +0300
+++ xorg-server/config/udev.c	2010-03-30 13:57:21.000000000 +0300
@@ -136,7 +136,7 @@
     }
     LogMessage(X_INFO, "config/udev: Adding input device %s (%s)\n",
                name, path);
-    rc = NewInputDeviceRequest(options, &attrs, &dev);
+    rc = NewInputDeviceRequest18(options, &attrs, &dev);
     if (rc != Success)
         goto unwind;
 
Index: xorg-server/hw/dmx/dmxinput.c
===================================================================
--- xorg-server.orig/hw/dmx/dmxinput.c	2010-03-30 13:57:21.000000000 +0300
+++ xorg-server/hw/dmx/dmxinput.c	2010-03-30 13:57:21.000000000 +0300
@@ -103,7 +103,13 @@
 }
 
 int
-NewInputDeviceRequest (InputOption *options, InputAttributes *attrs,
+NewInputDeviceRequest (InputOption *options, DeviceIntPtr *pdev)
+{
+    return BadRequest;
+}
+
+int
+NewInputDeviceRequest18 (InputOption *options, InputAttributes *attrs,
                        DeviceIntPtr *pdev)
 {
     return BadRequest;
Index: xorg-server/hw/kdrive/src/kinput.c
===================================================================
--- xorg-server.orig/hw/kdrive/src/kinput.c	2010-03-30 13:57:21.000000000 +0300
+++ xorg-server/hw/kdrive/src/kinput.c	2010-03-30 13:57:21.000000000 +0300
@@ -2249,7 +2249,12 @@
 }
 
 int
-NewInputDeviceRequest(InputOption *options, InputAttributes *attrs,
+NewInputDeviceRequest(InputOption *options, DeviceIntPtr *pdev)
+{
+    return NewInputDeviceRequest18(options, NULL, pdev);
+}
+int
+NewInputDeviceRequest18(InputOption *options, InputAttributes *attrs,
                       DeviceIntPtr *pdev)
 {
     InputOption *option = NULL;
Index: xorg-server/hw/xfree86/common/xf86Xinput.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86Xinput.c	2010-03-30 13:57:21.000000000 +0300
+++ xorg-server/hw/xfree86/common/xf86Xinput.c	2010-03-30 13:57:32.000000000 +0300
@@ -750,7 +750,13 @@
 }
 
 int
-NewInputDeviceRequest (InputOption *options, InputAttributes *attrs,
+NewInputDeviceRequest (InputOption *options, DeviceIntPtr *pdev)
+{
+    return NewInputDeviceRequest18(options, NULL, pdev);
+}
+
+int
+NewInputDeviceRequest18 (InputOption *options, InputAttributes *attrs,
                        DeviceIntPtr *pdev)
 {
     IDevRec *idev = NULL;
Index: xorg-server/hw/xquartz/darwinXinput.c
===================================================================
--- xorg-server.orig/hw/xquartz/darwinXinput.c	2010-03-30 13:57:21.000000000 +0300
+++ xorg-server/hw/xquartz/darwinXinput.c	2010-03-30 13:57:21.000000000 +0300
@@ -230,7 +230,14 @@
  *
  */
 int
-NewInputDeviceRequest(InputOption *options, InputAttributes *attrs,
+NewInputDeviceRequest(InputOption *options, DeviceIntPtr *pdev)
+{
+  DEBUG_LOG("NewInputDeviceRequest(%p, %p)\n", options, pdev);
+  return BadValue;
+}
+
+int
+NewInputDeviceRequest18(InputOption *options, InputAttributes *attrs,
                       DeviceIntPtr *pdev)
 {
   DEBUG_LOG("NewInputDeviceRequest(%p, %p)\n", options, pdev);
Index: xorg-server/include/input.h
===================================================================
--- xorg-server.orig/include/input.h	2010-03-30 13:57:21.000000000 +0300
+++ xorg-server/include/input.h	2010-03-30 13:57:21.000000000 +0300
@@ -530,6 +530,9 @@
 /* Implemented by the DDX. */
 extern _X_EXPORT int NewInputDeviceRequest(
     InputOption *options,
+    DeviceIntPtr *dev);
+extern int NewInputDeviceRequest18(
+    InputOption *options,
     InputAttributes *attrs,
     DeviceIntPtr *dev);
 extern  _X_EXPORT void DeleteInputDeviceRequest(
