Index: xorg-server-1.0.2/xkb/xkbInit.c
===================================================================
--- xorg-server-1.0.2.orig/xkb/xkbInit.c
+++ xorg-server-1.0.2/xkb/xkbInit.c
@@ -906,6 +906,14 @@
 extern unsigned char	XkbDfltAccessXOptions;
 
 int
+xkbInitGlobals()
+{
+    char * xkbpath= getenv("XKBPATH");
+    if (xkbpath!=NULL)
+	XkbBaseDirectory= xkbpath;
+}
+
+int
 XkbProcessArguments(int argc,char *argv[],int i)
 {
     if (strcmp(argv[i],"-kb")==0) {
Index: xorg-server-1.0.2/os/utils.c
===================================================================
--- xorg-server-1.0.2.orig/os/utils.c
+++ xorg-server-1.0.2/os/utils.c
@@ -712,6 +712,7 @@
  */
 void InitGlobals(void)
 {
+    xkbInitGlobals();
     ddxInitGlobals();
 }
 
