Index: xorg-server/xkb/xkbInit.c
===================================================================
--- xorg-server.orig/xkb/xkbInit.c	2006-11-13 19:59:23.000000000 +0100
+++ xorg-server/xkb/xkbInit.c	2006-11-26 02:00:19.000000000 +0100
@@ -841,6 +841,14 @@
 extern unsigned char	XkbDfltAccessXOptions;
 
 int
+xkbInitGlobals()
+{
+    char * xkbpath= getenv("XKBPATH");
+    if (xkbpath!=NULL)
+	XkbBaseDirectory= xkbpath;
+}
+
+int
 XkbProcessArguments(int argc,char *argv[],int i)
 {
     if (strcmp(argv[i],"-kb")==0) {
Index: xorg-server/os/utils.c
===================================================================
--- xorg-server.orig/os/utils.c	2006-11-13 19:59:23.000000000 +0100
+++ xorg-server/os/utils.c	2006-11-26 02:00:19.000000000 +0100
@@ -712,6 +712,7 @@
  */
 void InitGlobals(void)
 {
+    xkbInitGlobals();
     ddxInitGlobals();
 }
 
