Index: xorg-server-1.1.0/xkb/xkbInit.c
===================================================================
--- xorg-server-1.1.0.orig/xkb/xkbInit.c	2006-03-25 22:35:48.000000000 +0000
+++ xorg-server-1.1.0/xkb/xkbInit.c	2006-08-06 14:05:30.000000000 +0000
@@ -874,6 +874,14 @@
 extern unsigned char	XkbDfltAccessXOptions;
 
 int
+xkbInitGlobals()
+{
+    char * xkbpath= getenv("XKBPATH");
+    if (xkbpath!=NULL)
+	XkbBaseDirectory= xkbpath;
+}
+
+int
 XkbProcessArguments(int argc,char *argv[],int i)
 {
     if (strcmp(argv[i],"-kb")==0) {
Index: xorg-server-1.1.0/os/utils.c
===================================================================
--- xorg-server-1.1.0.orig/os/utils.c	2006-03-25 19:52:05.000000000 +0000
+++ xorg-server-1.1.0/os/utils.c	2006-08-06 14:05:30.000000000 +0000
@@ -715,6 +715,7 @@
  */
 void InitGlobals(void)
 {
+    xkbInitGlobals();
     ddxInitGlobals();
 }
 
