Index: xorg-server/hw/xfree86/common/xf86Xinput.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86Xinput.c	2010-03-29 19:57:48.000000000 +0300
+++ xorg-server/hw/xfree86/common/xf86Xinput.c	2010-03-29 19:59:19.000000000 +0300
@@ -827,13 +827,13 @@
     }
 
     if (!idev->driver || !idev->identifier) {
-        xf86Msg(X_ERROR, "No input driver/identifier specified (ignoring)\n");
+        xf86Msg(X_INFO, "No input driver/identifier specified (ignoring)\n");
         rval = BadRequest;
         goto unwind;
     }
 
     if (!idev->identifier) {
-        xf86Msg(X_ERROR, "No device identifier specified (ignoring)\n");
+        xf86Msg(X_INFO, "No device identifier specified (ignoring)\n");
         return BadMatch;
     }
 
