Index: xorg-server/hw/xfree86/common/xf86Config.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86Config.c	2007-11-19 20:36:12.000000000 -0500
+++ xorg-server/hw/xfree86/common/xf86Config.c	2007-11-19 20:36:13.000000000 -0500
@@ -1800,11 +1800,6 @@
     if (!servlayoutp)
 	return FALSE;
 
-    if (conf_screen == NULL) {
-	xf86ConfigError("No Screen sections present\n");
-	return FALSE;
-    }
-
     /*
      * which screen section is the active one?
      *
@@ -1892,6 +1887,12 @@
     XF86ConfAdaptorLinkPtr conf_adaptor;
     Bool defaultMonitor = FALSE;
 
+    if (!conf_screen) {
+        conf_screen = xnfcalloc(1, sizeof(XF86ConfScreenRec));
+        conf_screen->scrn_identifier = "Default Screen Section";
+        xf86Msg(X_DEFAULT, "No screen section available. Using defaults.\n");
+    }
+
     xf86Msg(from, "|-->Screen \"%s\" (%d)\n", conf_screen->scrn_identifier,
 	    scrnum);
     /*
Index: xorg-server/hw/xfree86/parser/Screen.c
===================================================================
--- xorg-server.orig/hw/xfree86/parser/Screen.c	2007-11-19 20:36:47.000000000 -0500
+++ xorg-server/hw/xfree86/parser/Screen.c	2007-11-19 20:36:57.000000000 -0500
@@ -498,12 +498,6 @@
 	XF86ConfDevicePtr device;
 	XF86ConfAdaptorLinkPtr adaptor;
 
-	if (!screen)
-	{
-		xf86validationError ("At least one Screen section is required.");
-		return (FALSE);
-	}
-
 	while (screen)
 	{
 		if (screen->scrn_obso_driver && !screen->scrn_identifier)
@@ -512,13 +506,7 @@
 		monitor = xf86findMonitor (screen->scrn_monitor_str, p->conf_monitor_lst);
 		if (screen->scrn_monitor_str)
 		{
-			if (!monitor)
-			{
-				xf86validationError (UNDEFINED_MONITOR_MSG,
-						 	screen->scrn_monitor_str, screen->scrn_identifier);
-				return (FALSE);
-			}
-			else
+			if (monitor)
 			{
 				screen->scrn_monitor = monitor;
 				if (!xf86validateMonitor(p, screen))
