Index: xorg-server/fb/fbpict.c
===================================================================
--- xorg-server.orig/fb/fbpict.c	2010-06-04 13:54:10.000000000 +1000
+++ xorg-server/fb/fbpict.c	2010-06-04 13:57:34.657009716 +1000
@@ -37,7 +37,7 @@
 #include "mipict.h"
 #include "fbpict.h"
 
-#define mod(a,b) ((b) == 1 ? 0 : (a) >= 0 ? (a) % (b) : (b) - (-a) % (b))
+#define mod(a,b) ((b) == 1 ? 0 : (a) >= 0 ? (a) % (b) : (b) - (-(a)) % (b))
 
 void
 fbWalkCompositeRegion (CARD8 op,
