Index: xorg-server-1.1.0/hw/xfree86/common/xf86Init.c
===================================================================
--- xorg-server-1.1.0.orig/hw/xfree86/common/xf86Init.c	2006-05-09 18:04:19.000000000 +0000
+++ xorg-server-1.1.0/hw/xfree86/common/xf86Init.c	2006-08-06 14:06:30.000000000 +0000
@@ -1905,7 +1905,11 @@
           FatalError("xf86RunVtInit: fork failed (%s)\n", strerror(errno));
           break;
       case 0:  /* child */
-          setuid(getuid());
+	  if (setuid(getuid()) == -1) {
+	      xf86Msg(X_ERROR, "xf86RunVtInit: setuid failed (%s)\n",
+			 strerror(errno));
+	      exit(255);
+	  }
           /* set stdin, stdout to the consoleFd */
           for (i = 0; i < 2; i++) {
             if (xf86Info.consoleFd != i) {
Index: xorg-server-1.1.0/hw/xfree86/os-support/shared/libc_wrapper.c
===================================================================
--- xorg-server-1.1.0.orig/hw/xfree86/os-support/shared/libc_wrapper.c	2006-03-25 19:52:04.000000000 +0000
+++ xorg-server-1.1.0/hw/xfree86/os-support/shared/libc_wrapper.c	2006-08-06 14:06:30.000000000 +0000
@@ -1270,7 +1270,10 @@
 #ifndef SELF_CONTAINED_WRAPPER
 	xf86DisableIO();
 #endif
-        setuid(getuid());
+        if (setuid(getuid()) == -1) {
+		ErrorF("xf86Execl: setuid() failed: %s\n", strerror(errno));
+		exit(255);
+	}
 #if !defined(SELF_CONTAINED_WRAPPER)
         /* set stdin, stdout to the consoleFD, and leave stderr alone */
         for (i = 0; i < 2; i++)
Index: xorg-server-1.1.0/hw/xfree86/parser/write.c
===================================================================
--- xorg-server-1.1.0.orig/hw/xfree86/parser/write.c	2005-07-03 07:01:37.000000000 +0000
+++ xorg-server-1.1.0/hw/xfree86/parser/write.c	2006-08-06 14:06:30.000000000 +0000
@@ -170,7 +170,10 @@
 					strerror(errno));
 			return 0;
 		case 0: /* child */
-			setuid(getuid());
+			if (setuid(getuid() == -1) 
+			    FatalError("xf86writeConfigFile(): "
+				"setuid failed(%s)\n", 
+				strerror(errno));
 			ret = doWriteConfigFile(filename, cptr);
 			exit(ret);
 			break;
Index: xorg-server-1.1.0/os/utils.c
===================================================================
--- xorg-server-1.1.0.orig/os/utils.c	2006-08-06 14:05:30.000000000 +0000
+++ xorg-server-1.1.0/os/utils.c	2006-08-06 14:06:30.000000000 +0000
@@ -1722,8 +1722,10 @@
     case -1:	/* error */
 	p = -1;
     case 0:	/* child */
-	setgid(getgid());
-	setuid(getuid());
+	if (setgid(getgid()) == -1)
+	    _exit(127);
+	if (setuid(getuid()) == -1)
+	    _exit(127);
 	execl("/bin/sh", "sh", "-c", command, (char *)NULL);
 	_exit(127);
     default:	/* parent */
@@ -1774,8 +1776,10 @@
 	xfree(cur);
 	return NULL;
     case 0:	/* child */
-	setgid(getgid());
-	setuid(getuid());
+	if (setgid(getgid()) == -1)
+	    _exit(127);
+	if (setuid(getuid()) == -1)
+	    _exit(127);
 	if (*type == 'r') {
 	    if (pdes[1] != 1) {
 		/* stdout */
@@ -1849,8 +1853,10 @@
 	xfree(cur);
 	return NULL;
     case 0:	/* child */
-	setgid(getgid());
-	setuid(getuid());
+	if (setgid(getgid()) == -1)
+	    _exit(127);
+	if (setuid(getuid()) == -1)
+	    _exit(127);
 	if (*type == 'r') {
 	    if (pdes[1] != 1) {
 		/* stdout */
