Index: xorg-server/hw/xfree86/os-support/linux/lnx_video.c
===================================================================
--- xorg-server.orig/hw/xfree86/os-support/linux/lnx_video.c	2006-11-13 19:59:23.000000000 +0100
+++ xorg-server/hw/xfree86/os-support/linux/lnx_video.c	2006-11-26 02:06:51.000000000 +0100
@@ -567,7 +567,7 @@
 #endif
 	}
 	close(fd);
-#elif !defined(__mc68000__) && !defined(__sparc__) && !defined(__mips__) && !defined(__sh__) && !defined(__hppa__)
+#elif !defined(__mc68000__) && !defined(__sparc__) && !defined(__mips__) && !defined(__sh__) && !defined(__hppa__) && !defined(__s390__)
         if (ioperm(0, 1024, 1) || iopl(3)) {
                 if (errno == ENODEV)
                         ErrorF("xf86EnableIOPorts: no I/O ports found\n");
Index: xorg-server/hw/xfree86/common/compiler.h
===================================================================
--- xorg-server.orig/hw/xfree86/common/compiler.h	2006-11-13 19:59:23.000000000 +0100
+++ xorg-server/hw/xfree86/common/compiler.h	2006-11-26 02:06:51.000000000 +0100
@@ -1365,7 +1365,7 @@
 #    define write_mem_barrier()   /* NOP */
 
 #    if !defined(__SUNPRO_C)
-#    if !defined(FAKEIT) && !defined(__mc68000__) && !defined(__arm__) && !defined(__sh__) && !defined(__hppa__)
+#    if !defined(FAKEIT) && !defined(__mc68000__) && !defined(__arm__) && !defined(__sh__) && !defined(__hppa__) && !defined(__s390__)
 #     ifdef GCCUSESGAS
 
 /*
