Index: xorg-server/GL/mesa/main/Makefile.am
===================================================================
--- xorg-server.orig/GL/mesa/main/Makefile.am	2006-09-28 21:51:09.000000000 -0400
+++ xorg-server/GL/mesa/main/Makefile.am	2006-09-28 22:33:17.000000000 -0400
@@ -29,6 +29,7 @@
                     api_loopback.c \
                     api_noop.c \
                     api_validate.c \
+					arrayobj.c \
                     attrib.c \
                     blend.c \
                     bufferobj.c \
@@ -67,6 +68,7 @@
                     points.c \
                     polygon.c \
                     rastpos.c \
+					rbadaptors.c \
                     renderbuffer.c \
                     state.c \
                     stencil.c \
