Index: xorg-server/GL/glx/Makefile.am
===================================================================
--- xorg-server.orig/GL/glx/Makefile.am	2007-08-16 22:36:59.000000000 -0400
+++ xorg-server/GL/glx/Makefile.am	2007-08-16 22:47:32.000000000 -0400
@@ -14,7 +14,8 @@
 	-I@MESA_SOURCE@/src/mesa/glapi \
 	-I@MESA_SOURCE@/src/mesa/main \
 	-DXFree86Server \
-	@GLX_DEFINES@
+	@GLX_DEFINES@ \
+	@GLX_ARCH_DEFINES@
 
 # none yet
 #sdk_HEADERS =
Index: xorg-server/configure.ac
===================================================================
--- xorg-server.orig/configure.ac	2007-08-16 22:46:38.000000000 -0400
+++ xorg-server/configure.ac	2007-08-16 22:47:32.000000000 -0400
@@ -276,6 +276,7 @@
 	case $host_os in
 		*netbsd*)	AC_DEFINE(USE_ALPHA_PIO, 1, [NetBSD PIO alpha IO]) ;;
 	esac
+	GLX_ARCH_DEFINES="-D__GLX_ALIGN64 -mieee"
 	;;
   arm*)
 	ARM_VIDEO=yes
@@ -305,6 +306,7 @@
 	xorg_loader_sparcmuldiv="yes"
 	SPARC64_VIDEO=yes
 	BSD_ARCH_SOURCES="sparc64_video.c ioperm_noop.c"
+	GLX_ARCH_DEFINES="-D__GLX_ALIGN64"
 	;;
   x86_64*|amd64*)
   	use_x86_asm="yes"
@@ -319,8 +321,16 @@
 				SYS_LIBS=-lamd64
 				;;
 	esac
+	GLX_ARCH_DEFINES="-D__GLX_ALIGN64"
+	;;
+  ia64*)
+  	GLX_ARCH_DEFINES="-D__GLX_ALIGN64"
+	;;
+  s390*)
+  	GLX_ARCH_DEFINES="-D__GLX_ALIGN64"
 	;;
 esac
+AC_SUBST(GLX_ARCH_DEFINES)
 
 dnl BSD *_video.c selection
 AM_CONDITIONAL(ALPHA_VIDEO, [test "x$ALPHA_VIDEO" = xyes])
Index: xorg-server/hw/dmx/glxProxy/Makefile.am
===================================================================
--- xorg-server.orig/hw/dmx/glxProxy/Makefile.am	2007-08-04 18:25:15.000000000 -0400
+++ xorg-server/hw/dmx/glxProxy/Makefile.am	2007-08-16 22:47:32.000000000 -0400
@@ -32,6 +32,7 @@
                         unpack.h
 
 AM_CFLAGS = \
+	    @GLX_ARCH_DEFINES@ \
             $(DIX_CFLAGS) \
             -I$(top_srcdir)/hw/dmx \
             -I$(top_srcdir)/include \
