Index: xorg-server/configure.ac
===================================================================
--- xorg-server.orig/configure.ac	2006-12-02 11:17:18.000000000 +0100
+++ xorg-server/configure.ac	2006-12-02 11:17:45.000000000 +0100
@@ -1512,7 +1512,11 @@
 
     AC_CHECK_HEADERS([sys/vm86.h sys/io.h])
     if test "$ac_cv_header_sys_vm86_h" = yes; then
-        AC_DEFINE(KDRIVEVESA, 1, [Build VESA-based kdrive servers])
+    	case $host_os in
+		kfreebsd*-gnu)	kdrivevesa=no ;;
+		*) 	AC_DEFINE(KDRIVEVESA, 1, [Build VESA-based kdrive servers])
+			kdrivevesa=yes;;
+	esac
     fi
 
     AC_CHECK_HEADERS([linux/fb.h])
@@ -1562,7 +1566,7 @@
 AC_SUBST(KDRIVE_LIBS)
 AM_CONDITIONAL(TSLIB, [test "x$HAVE_TSLIB" = xyes])
 AM_CONDITIONAL(H3600_TS, false)
-AM_CONDITIONAL(KDRIVEVESA, [test x"$ac_cv_header_sys_vm86_h" = xyes])
+AM_CONDITIONAL(KDRIVEVESA, [test x"$kdrivevesa" = xyes])
 AM_CONDITIONAL(KDRIVEFBDEV, [test x"$ac_cv_header_linux_fb_h" = xyes])
 
 # Xephyr needs nanosleep() which is in librt on Solaris
Index: xorg-server/hw/kdrive/linux/agp.c
===================================================================
--- xorg-server.orig/hw/kdrive/linux/agp.c	2006-12-02 11:14:28.000000000 +0100
+++ xorg-server/hw/kdrive/linux/agp.c	2006-12-02 11:17:45.000000000 +0100
@@ -65,7 +65,7 @@
 
 #include <linux/agpgart.h>
 
-#elif defined(__FreeBSD__)
+#elif defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
 #include <sys/ioctl.h>
 #include <sys/agpio.h>
 #endif
