Index: xorg-server/hw/kdrive/linux/agp.c
===================================================================
--- xorg-server.orig/hw/kdrive/linux/agp.c	2007-05-23 17:22:18.000000000 +0200
+++ xorg-server/hw/kdrive/linux/agp.c	2007-05-23 18:06:59.000000000 +0200
@@ -65,7 +65,7 @@
 
 #include <linux/agpgart.h>
 
-#elif defined(__FreeBSD__)
+#elif defined(__FreeBSD__) || defined(__FreeBSD_kernel__)
 #include <sys/ioctl.h>
 #include <sys/agpio.h>
 #endif
