Index: xorg-server/configure
===================================================================
--- xorg-server.orig/configure	2006-09-28 22:37:18.000000000 -0400
+++ xorg-server/configure	2006-09-28 23:09:07.000000000 -0400
@@ -883,6 +883,7 @@
 FBDEVHW_FALSE
 FREEBSD_KLDLOAD_TRUE
 FREEBSD_KLDLOAD_FALSE
+GLX_ARCH_DEFINES
 ALPHA_VIDEO_TRUE
 ALPHA_VIDEO_FALSE
 ARM_VIDEO_TRUE
@@ -987,6 +988,8 @@
 XPRINT_FALSE
 XINPUT_TRUE
 XINPUT_FALSE
+MISCUTILS_TRUE
+MISCUTILS_FALSE
 XDMCP_CFLAGS
 XDMCP_LIBS
 XDMCP_TRUE
@@ -997,6 +1000,7 @@
 USE_RGB_BUILTIN_FALSE
 COMPILEDDEFAULTFONTPATH
 RGB_DB
+SERVERCONFIGdir
 DRI_DRIVER_PATH
 VENDOR_STRING
 VENDOR_STRING_SHORT
@@ -1147,6 +1151,8 @@
 XWIN_XV_FALSE
 KDRIVE_TRUE
 KDRIVE_FALSE
+KDRIVELINUX_TRUE
+KDRIVELINUX_FALSE
 XEPHYR_CFLAGS
 XEPHYR_LIBS
 KDRIVE_INCS
@@ -1907,6 +1913,7 @@
   --disable-xf86bigfont   Build XF86 Big Font extension (default: enabled)
   --disable-dpms          Build DPMS extension (default: enabled)
   --disable-xinput        Build XInput Extension (default: enabled)
+  --enable-misc-utils     Build misc utilities (default: enabled)
   --enable-xorg           Build Xorg server (default: auto)
   --enable-dmx            Build DMX server (default: auto)
   --enable-xvfb           Build Xvfb server (default: yes)
@@ -1971,6 +1978,8 @@
   --with-xkb-output=PATH  Path to XKB output dir (default:
                           ${datadir}/X11/xkb/compiled)
   --with-rgb-path=PATH    Path to RGB database (default: ${datadir}/X11/rgb)
+  --with-serverconfig-path=PATH
+                          Path to server config (default: ${libdir}/xserver)
   --with-dri-driver-path=PATH
                           Path to DRI drivers (default: ${libdir}/dri)
    --with-freetype-config=PROG
@@ -5103,7 +5112,7 @@
   ;;
 *-*-irix6*)
   # Find out which ABI we are using.
-  echo '#line 5106 "configure"' > conftest.$ac_ext
+  echo '#line 5115 "configure"' > conftest.$ac_ext
   if { (eval echo "$as_me:$LINENO: \"$ac_compile\"") >&5
   (eval $ac_compile) 2>&5
   ac_status=$?
@@ -8357,11 +8366,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:8360: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:8369: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>conftest.err)
    ac_status=$?
    cat conftest.err >&5
-   echo "$as_me:8364: \$? = $ac_status" >&5
+   echo "$as_me:8373: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s "$ac_outfile"; then
      # The compiler can only warn and ignore the option if not recognized
      # So say no if there are warnings other than the usual output.
@@ -8625,11 +8634,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:8628: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:8637: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>conftest.err)
    ac_status=$?
    cat conftest.err >&5
-   echo "$as_me:8632: \$? = $ac_status" >&5
+   echo "$as_me:8641: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s "$ac_outfile"; then
      # The compiler can only warn and ignore the option if not recognized
      # So say no if there are warnings other than the usual output.
@@ -8729,11 +8738,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:8732: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:8741: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>out/conftest.err)
    ac_status=$?
    cat out/conftest.err >&5
-   echo "$as_me:8736: \$? = $ac_status" >&5
+   echo "$as_me:8745: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s out/conftest2.$ac_objext
    then
      # The compiler can only warn and ignore the option if not recognized
@@ -11170,7 +11179,7 @@
   lt_dlunknown=0; lt_dlno_uscore=1; lt_dlneed_uscore=2
   lt_status=$lt_dlunknown
   cat > conftest.$ac_ext <<EOF
-#line 11173 "configure"
+#line 11182 "configure"
 #include "confdefs.h"
 
 #if HAVE_DLFCN_H
@@ -11270,7 +11279,7 @@
   lt_dlunknown=0; lt_dlno_uscore=1; lt_dlneed_uscore=2
   lt_status=$lt_dlunknown
   cat > conftest.$ac_ext <<EOF
-#line 11273 "configure"
+#line 11282 "configure"
 #include "confdefs.h"
 
 #if HAVE_DLFCN_H
@@ -13638,11 +13647,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:13641: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:13650: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>conftest.err)
    ac_status=$?
    cat conftest.err >&5
-   echo "$as_me:13645: \$? = $ac_status" >&5
+   echo "$as_me:13654: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s "$ac_outfile"; then
      # The compiler can only warn and ignore the option if not recognized
      # So say no if there are warnings other than the usual output.
@@ -13742,11 +13751,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:13745: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:13754: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>out/conftest.err)
    ac_status=$?
    cat out/conftest.err >&5
-   echo "$as_me:13749: \$? = $ac_status" >&5
+   echo "$as_me:13758: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s out/conftest2.$ac_objext
    then
      # The compiler can only warn and ignore the option if not recognized
@@ -15303,11 +15312,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:15306: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:15315: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>conftest.err)
    ac_status=$?
    cat conftest.err >&5
-   echo "$as_me:15310: \$? = $ac_status" >&5
+   echo "$as_me:15319: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s "$ac_outfile"; then
      # The compiler can only warn and ignore the option if not recognized
      # So say no if there are warnings other than the usual output.
@@ -15407,11 +15416,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:15410: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:15419: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>out/conftest.err)
    ac_status=$?
    cat out/conftest.err >&5
-   echo "$as_me:15414: \$? = $ac_status" >&5
+   echo "$as_me:15423: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s out/conftest2.$ac_objext
    then
      # The compiler can only warn and ignore the option if not recognized
@@ -17626,11 +17635,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:17629: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:17638: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>conftest.err)
    ac_status=$?
    cat conftest.err >&5
-   echo "$as_me:17633: \$? = $ac_status" >&5
+   echo "$as_me:17642: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s "$ac_outfile"; then
      # The compiler can only warn and ignore the option if not recognized
      # So say no if there are warnings other than the usual output.
@@ -17894,11 +17903,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:17897: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:17906: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>conftest.err)
    ac_status=$?
    cat conftest.err >&5
-   echo "$as_me:17901: \$? = $ac_status" >&5
+   echo "$as_me:17910: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s "$ac_outfile"; then
      # The compiler can only warn and ignore the option if not recognized
      # So say no if there are warnings other than the usual output.
@@ -17998,11 +18007,11 @@
    -e 's:.*FLAGS}\{0,1\} :&$lt_compiler_flag :; t' \
    -e 's: [^ ]*conftest\.: $lt_compiler_flag&:; t' \
    -e 's:$: $lt_compiler_flag:'`
-   (eval echo "\"\$as_me:18001: $lt_compile\"" >&5)
+   (eval echo "\"\$as_me:18010: $lt_compile\"" >&5)
    (eval "$lt_compile" 2>out/conftest.err)
    ac_status=$?
    cat out/conftest.err >&5
-   echo "$as_me:18005: \$? = $ac_status" >&5
+   echo "$as_me:18014: \$? = $ac_status" >&5
    if (exit $ac_status) && test -s out/conftest2.$ac_objext
    then
      # The compiler can only warn and ignore the option if not recognized
@@ -26046,6 +26055,7 @@
 _ACEOF
  ;;
 	esac
+	GLX_ARCH_DEFINES="-D__GLX_ALIGN64 -mieee"
 	;;
   arm*)
 	ARM_VIDEO=yes
@@ -26081,6 +26091,7 @@
 	xorg_loader_sparcmuldiv="yes"
 	SPARC64_VIDEO=yes
 	BSD_ARCH_SOURCES="sparc64_video.c ioperm_noop.c"
+	GLX_ARCH_DEFINES="-D__GLX_ALIGN64"
 	;;
   x86_64*|amd64*)
   	use_x86_asm="yes"
@@ -26104,11 +26115,19 @@
 				SYS_LIBS=-lamd64
 				;;
 	esac
+	GLX_ARCH_DEFINES="-D__GLX_ALIGN64"
+	;;
+  ia64*)
+  	GLX_ARCH_DEFINES="-D__GLX_ALIGN64"
+	;;
+  s390*)
+  	GLX_ARCH_DEFINES="-D__GLX_ALIGN64"
 	;;
 esac
 
 
 
+
 if test "x$ALPHA_VIDEO" = xyes; then
   ALPHA_VIDEO_TRUE=
   ALPHA_VIDEO_FALSE='#'
@@ -26486,6 +26505,14 @@
 fi
 
 
+# Check whether --with-serverconfig-path was given.
+if test "${with_serverconfig_path+set}" = set; then
+  withval=$with_serverconfig_path;  SERVERCONFIG="$withval"
+else
+   SERVERCONFIG="${libdir}/xserver"
+fi
+
+
 # Check whether --with-dri-driver-path was given.
 if test "${with_dri_driver_path+set}" = set; then
   withval=$with_dri_driver_path;  DRI_DRIVER_PATH="$withval"
@@ -26716,6 +26743,13 @@
   XINPUT=yes
 fi
 
+# Check whether --enable-misc-utils was given.
+if test "${enable_misc_utils+set}" = set; then
+  enableval=$enable_misc_utils; MISCUTILS=$enableval
+else
+  MISCUTILS=yes
+fi
+
 
 # Check whether --enable-xorg was given.
 if test "${enable_xorg+set}" = set; then
@@ -29418,6 +29452,17 @@
 fi
 
 
+
+if test "x$MISCUTILS" = xyes; then
+  MISCUTILS_TRUE=
+  MISCUTILS_FALSE='#'
+else
+  MISCUTILS_TRUE='#'
+  MISCUTILS_FALSE=
+fi
+
+
+
 cat >>confdefs.h <<\_ACEOF
 #define SHAPE 1
 _ACEOF
@@ -29833,6 +29878,22 @@
   exec_prefix_NONE=
   test "x$prefix" = xNONE && prefix_NONE=yes && prefix=$ac_default_prefix
   test "x$exec_prefix" = xNONE && exec_prefix_NONE=yes && exec_prefix=$prefix
+  eval ac_define_dir="\"$SERVERCONFIG\""
+  SERVERCONFIGdir="$ac_define_dir"
+
+
+cat >>confdefs.h <<_ACEOF
+#define SERVERCONFIGdir "$ac_define_dir"
+_ACEOF
+
+  test "$prefix_NONE" && prefix=NONE
+  test "$exec_prefix_NONE" && exec_prefix=NONE
+
+
+  prefix_NONE=
+  exec_prefix_NONE=
+  test "x$prefix" = xNONE && prefix_NONE=yes && prefix=$ac_default_prefix
+  test "x$exec_prefix" = xNONE && exec_prefix_NONE=yes && exec_prefix=$prefix
   eval ac_define_dir="\"$DRI_DRIVER_PATH\""
   DRI_DRIVER_PATH="$ac_define_dir"
 
@@ -34040,7 +34101,33 @@
     #    $MIEXT_SHADOW_LIB $XPSTUBS_LIB"
     KDRIVE_PURE_LIBS="$FB_LIB $MI_LIB $FIXES_LIB $XEXT_LIB $DBE_LIB $XTRAP_LIB $RECORD_LIB $GLX_LIBS $RENDER_LIB $RANDR_LIB $DAMAGE_LIB $MIEXT_DAMAGE_LIB $MIEXT_SHADOW_LIB $XI_LIB $XKB_LIB $XKB_STUB_LIB $COMPOSITE_LIB $XPSTUBS_LIB $OS_LIB"
     KDRIVE_LIB='$(top_builddir)/hw/kdrive/src/libkdrive.a'
-    KDRIVE_OS_LIB='$(top_builddir)/hw/kdrive/linux/liblinux.a'
+    case $host_os in
+	*linux*)
+	    KDRIVE_OS_LIB='$(top_builddir)/hw/kdrive/linux/liblinux.a'
+
+
+if true; then
+  KDRIVELINUX_TRUE=
+  KDRIVELINUX_FALSE='#'
+else
+  KDRIVELINUX_TRUE='#'
+  KDRIVELINUX_FALSE=
+fi
+
+	    ;;
+	*)
+
+
+if false; then
+  KDRIVELINUX_TRUE=
+  KDRIVELINUX_FALSE='#'
+else
+  KDRIVELINUX_TRUE='#'
+  KDRIVELINUX_FALSE=
+fi
+
+	    ;;
+    esac
     KDRIVE_STUB_LIB='$(top_builddir)/hw/kdrive/src/libkdrivestubs.a'
     KDRIVE_LIBS="$DIX_LIB $KDRIVE_LIB $KDRIVE_OS_LIB $KDRIVE_PURE_LIBS $KDRIVE_STUB_LIB"
 
@@ -35537,6 +35624,13 @@
 Usually this means the macro was only invoked conditionally." >&2;}
    { (exit 1); exit 1; }; }
 fi
+if test -z "${MISCUTILS_TRUE}" && test -z "${MISCUTILS_FALSE}"; then
+  { { echo "$as_me:$LINENO: error: conditional \"MISCUTILS\" was never defined.
+Usually this means the macro was only invoked conditionally." >&5
+echo "$as_me: error: conditional \"MISCUTILS\" was never defined.
+Usually this means the macro was only invoked conditionally." >&2;}
+   { (exit 1); exit 1; }; }
+fi
 if test -z "${XDMCP_TRUE}" && test -z "${XDMCP_FALSE}"; then
   { { echo "$as_me:$LINENO: error: conditional \"XDMCP\" was never defined.
 Usually this means the macro was only invoked conditionally." >&5
@@ -35817,6 +35911,20 @@
 Usually this means the macro was only invoked conditionally." >&2;}
    { (exit 1); exit 1; }; }
 fi
+if test -z "${KDRIVELINUX_TRUE}" && test -z "${KDRIVELINUX_FALSE}"; then
+  { { echo "$as_me:$LINENO: error: conditional \"KDRIVELINUX\" was never defined.
+Usually this means the macro was only invoked conditionally." >&5
+echo "$as_me: error: conditional \"KDRIVELINUX\" was never defined.
+Usually this means the macro was only invoked conditionally." >&2;}
+   { (exit 1); exit 1; }; }
+fi
+if test -z "${KDRIVELINUX_TRUE}" && test -z "${KDRIVELINUX_FALSE}"; then
+  { { echo "$as_me:$LINENO: error: conditional \"KDRIVELINUX\" was never defined.
+Usually this means the macro was only invoked conditionally." >&5
+echo "$as_me: error: conditional \"KDRIVELINUX\" was never defined.
+Usually this means the macro was only invoked conditionally." >&2;}
+   { (exit 1); exit 1; }; }
+fi
 if test -z "${TSLIB_TRUE}" && test -z "${TSLIB_FALSE}"; then
   { { echo "$as_me:$LINENO: error: conditional \"TSLIB\" was never defined.
 Usually this means the macro was only invoked conditionally." >&5
@@ -36761,6 +36869,7 @@
 FBDEVHW_FALSE!$FBDEVHW_FALSE$ac_delim
 FREEBSD_KLDLOAD_TRUE!$FREEBSD_KLDLOAD_TRUE$ac_delim
 FREEBSD_KLDLOAD_FALSE!$FREEBSD_KLDLOAD_FALSE$ac_delim
+GLX_ARCH_DEFINES!$GLX_ARCH_DEFINES$ac_delim
 ALPHA_VIDEO_TRUE!$ALPHA_VIDEO_TRUE$ac_delim
 ALPHA_VIDEO_FALSE!$ALPHA_VIDEO_FALSE$ac_delim
 ARM_VIDEO_TRUE!$ARM_VIDEO_TRUE$ac_delim
@@ -36825,7 +36934,6 @@
 AIGLX_TRUE!$AIGLX_TRUE$ac_delim
 AIGLX_FALSE!$AIGLX_FALSE$ac_delim
 GLX_DEFINES!$GLX_DEFINES$ac_delim
-LBXUTIL_TEST_CFLAGS!$LBXUTIL_TEST_CFLAGS$ac_delim
 _ACEOF
 
   if test `sed -n "s/.*$ac_delim\$/X/p" conf$$subs.sed | grep -c X` = 97; then
@@ -36867,6 +36975,7 @@
 ac_delim='%!_!# '
 for ac_last_try in false false false false false :; do
   cat >conf$$subs.sed <<_ACEOF
+LBXUTIL_TEST_CFLAGS!$LBXUTIL_TEST_CFLAGS$ac_delim
 LBXUTIL_TEST_LIBS!$LBXUTIL_TEST_LIBS$ac_delim
 LBX_TRUE!$LBX_TRUE$ac_delim
 LBX_FALSE!$LBX_FALSE$ac_delim
@@ -36906,6 +37015,8 @@
 XPRINT_FALSE!$XPRINT_FALSE$ac_delim
 XINPUT_TRUE!$XINPUT_TRUE$ac_delim
 XINPUT_FALSE!$XINPUT_FALSE$ac_delim
+MISCUTILS_TRUE!$MISCUTILS_TRUE$ac_delim
+MISCUTILS_FALSE!$MISCUTILS_FALSE$ac_delim
 XDMCP_CFLAGS!$XDMCP_CFLAGS$ac_delim
 XDMCP_LIBS!$XDMCP_LIBS$ac_delim
 XDMCP_TRUE!$XDMCP_TRUE$ac_delim
@@ -36916,6 +37027,7 @@
 USE_RGB_BUILTIN_FALSE!$USE_RGB_BUILTIN_FALSE$ac_delim
 COMPILEDDEFAULTFONTPATH!$COMPILEDDEFAULTFONTPATH$ac_delim
 RGB_DB!$RGB_DB$ac_delim
+SERVERCONFIGdir!$SERVERCONFIGdir$ac_delim
 DRI_DRIVER_PATH!$DRI_DRIVER_PATH$ac_delim
 VENDOR_STRING!$VENDOR_STRING$ac_delim
 VENDOR_STRING_SHORT!$VENDOR_STRING_SHORT$ac_delim
@@ -36960,10 +37072,6 @@
 XNEST_TRUE!$XNEST_TRUE$ac_delim
 XNEST_FALSE!$XNEST_FALSE$ac_delim
 XNEST_LIBS!$XNEST_LIBS$ac_delim
-XGLMODULES_CFLAGS!$XGLMODULES_CFLAGS$ac_delim
-XGLMODULES_LIBS!$XGLMODULES_LIBS$ac_delim
-XGL_TRUE!$XGL_TRUE$ac_delim
-XGL_FALSE!$XGL_FALSE$ac_delim
 _ACEOF
 
   if test `sed -n "s/.*$ac_delim\$/X/p" conf$$subs.sed | grep -c X` = 97; then
@@ -37005,6 +37113,10 @@
 ac_delim='%!_!# '
 for ac_last_try in false false false false false :; do
   cat >conf$$subs.sed <<_ACEOF
+XGLMODULES_CFLAGS!$XGLMODULES_CFLAGS$ac_delim
+XGLMODULES_LIBS!$XGLMODULES_LIBS$ac_delim
+XGL_TRUE!$XGL_TRUE$ac_delim
+XGL_FALSE!$XGL_FALSE$ac_delim
 XGL_LIBS!$XGL_LIBS$ac_delim
 xglmoduledir!$xglmoduledir$ac_delim
 XGL_MODULE_PATH!$XGL_MODULE_PATH$ac_delim
@@ -37098,10 +37210,6 @@
 XWIN_GLX_WINDOWS_TRUE!$XWIN_GLX_WINDOWS_TRUE$ac_delim
 XWIN_GLX_WINDOWS_FALSE!$XWIN_GLX_WINDOWS_FALSE$ac_delim
 XWIN_NATIVEGDI_TRUE!$XWIN_NATIVEGDI_TRUE$ac_delim
-XWIN_NATIVEGDI_FALSE!$XWIN_NATIVEGDI_FALSE$ac_delim
-XWIN_PRIMARYFB_TRUE!$XWIN_PRIMARYFB_TRUE$ac_delim
-XWIN_PRIMARYFB_FALSE!$XWIN_PRIMARYFB_FALSE$ac_delim
-XWIN_RANDR_TRUE!$XWIN_RANDR_TRUE$ac_delim
 _ACEOF
 
   if test `sed -n "s/.*$ac_delim\$/X/p" conf$$subs.sed | grep -c X` = 97; then
@@ -37143,11 +37251,17 @@
 ac_delim='%!_!# '
 for ac_last_try in false false false false false :; do
   cat >conf$$subs.sed <<_ACEOF
+XWIN_NATIVEGDI_FALSE!$XWIN_NATIVEGDI_FALSE$ac_delim
+XWIN_PRIMARYFB_TRUE!$XWIN_PRIMARYFB_TRUE$ac_delim
+XWIN_PRIMARYFB_FALSE!$XWIN_PRIMARYFB_FALSE$ac_delim
+XWIN_RANDR_TRUE!$XWIN_RANDR_TRUE$ac_delim
 XWIN_RANDR_FALSE!$XWIN_RANDR_FALSE$ac_delim
 XWIN_XV_TRUE!$XWIN_XV_TRUE$ac_delim
 XWIN_XV_FALSE!$XWIN_XV_FALSE$ac_delim
 KDRIVE_TRUE!$KDRIVE_TRUE$ac_delim
 KDRIVE_FALSE!$KDRIVE_FALSE$ac_delim
+KDRIVELINUX_TRUE!$KDRIVELINUX_TRUE$ac_delim
+KDRIVELINUX_FALSE!$KDRIVELINUX_FALSE$ac_delim
 XEPHYR_CFLAGS!$XEPHYR_CFLAGS$ac_delim
 XEPHYR_LIBS!$XEPHYR_LIBS$ac_delim
 KDRIVE_INCS!$KDRIVE_INCS$ac_delim
@@ -37214,7 +37328,7 @@
 LTLIBOBJS!$LTLIBOBJS$ac_delim
 _ACEOF
 
-  if test `sed -n "s/.*$ac_delim\$/X/p" conf$$subs.sed | grep -c X` = 69; then
+  if test `sed -n "s/.*$ac_delim\$/X/p" conf$$subs.sed | grep -c X` = 75; then
     break
   elif $ac_last_try; then
     { { echo "$as_me:$LINENO: error: could not make $CONFIG_STATUS" >&5
Index: xorg-server/debian/xserver-xorg-core.install
===================================================================
--- xorg-server.orig/debian/xserver-xorg-core.install	2006-09-28 22:37:18.000000000 -0400
+++ xorg-server/debian/xserver-xorg-core.install	2006-09-28 23:08:47.000000000 -0400
@@ -3,6 +3,7 @@
 usr/lib/xorg/modules/extensions/*.so
 usr/lib/xorg/modules/fonts/*
 usr/lib/xorg/modules/linux/*.so
+usr/lib/xorg/modules/freebsd/*.so
 usr/lib/xorg/modules/multimedia/*.so
 usr/bin/Xorg
 usr/share/man/man1/Xserver.1
