Index: GL/mesa/main/Makefile.in
===================================================================
--- GL/mesa/main/Makefile.in.orig	2006-09-28 22:37:17.000000000 -0400
+++ GL/mesa/main/Makefile.in	2006-09-28 23:09:20.000000000 -0400
@@ -57,17 +57,18 @@
 libmain_la_LIBADD =
 @NEED_VSNPRINTF_TRUE@am__objects_1 = vsnprintf.lo
 nodist_libmain_la_OBJECTS = accum.lo api_arrayelt.lo api_loopback.lo \
-	api_noop.lo api_validate.lo attrib.lo blend.lo bufferobj.lo \
-	buffers.lo clip.lo colortab.lo context.lo convolve.lo debug.lo \
-	depth.lo depthstencil.lo dlist.lo drawpix.lo enable.lo \
-	enums.lo eval.lo execmem.lo extensions.lo fbobject.lo \
-	feedback.lo fog.lo framebuffer.lo get.lo getstring.lo hash.lo \
-	hint.lo histogram.lo image.lo imports.lo light.lo lines.lo \
-	matrix.lo mm.lo occlude.lo pixel.lo points.lo polygon.lo \
-	rastpos.lo renderbuffer.lo state.lo stencil.lo texcompress.lo \
-	texcompress_fxt1.lo texcompress_s3tc.lo texenvprogram.lo \
-	texformat.lo teximage.lo texobj.lo texrender.lo texstate.lo \
-	texstore.lo varray.lo $(am__objects_1) vtxfmt.lo
+	api_noop.lo api_validate.lo arrayobj.lo attrib.lo blend.lo \
+	bufferobj.lo buffers.lo clip.lo colortab.lo context.lo \
+	convolve.lo debug.lo depth.lo depthstencil.lo dlist.lo \
+	drawpix.lo enable.lo enums.lo eval.lo execmem.lo extensions.lo \
+	fbobject.lo feedback.lo fog.lo framebuffer.lo get.lo \
+	getstring.lo hash.lo hint.lo histogram.lo image.lo imports.lo \
+	light.lo lines.lo matrix.lo mm.lo occlude.lo pixel.lo \
+	points.lo polygon.lo rastpos.lo rbadaptors.lo renderbuffer.lo \
+	state.lo stencil.lo texcompress.lo texcompress_fxt1.lo \
+	texcompress_s3tc.lo texenvprogram.lo texformat.lo teximage.lo \
+	texobj.lo texrender.lo texstate.lo texstore.lo varray.lo \
+	$(am__objects_1) vtxfmt.lo
 libmain_la_OBJECTS = $(nodist_libmain_la_OBJECTS)
 DEFAULT_INCLUDES = -I. -I$(srcdir) -I$(top_builddir)/include -I$(top_builddir)/include -I$(top_builddir)/include -I$(top_builddir)/include -I$(top_builddir)/include -I$(top_builddir)/include -I$(top_builddir)/include -I$(top_builddir)/include
 depcomp = $(SHELL) $(top_srcdir)/depcomp
@@ -201,6 +202,7 @@
 FREETYPE_CFLAGS = @FREETYPE_CFLAGS@
 FREETYPE_LIBS = @FREETYPE_LIBS@
 FREETYPE_REQUIRES = @FREETYPE_REQUIRES@
+GLX_ARCH_DEFINES = @GLX_ARCH_DEFINES@
 GLX_DEFINES = @GLX_DEFINES@
 GLX_FALSE = @GLX_FALSE@
 GLX_TRUE = @GLX_TRUE@
@@ -227,6 +229,8 @@
 INT10_X86EMU_TRUE = @INT10_X86EMU_TRUE@
 KDRIVEFBDEV_FALSE = @KDRIVEFBDEV_FALSE@
 KDRIVEFBDEV_TRUE = @KDRIVEFBDEV_TRUE@
+KDRIVELINUX_FALSE = @KDRIVELINUX_FALSE@
+KDRIVELINUX_TRUE = @KDRIVELINUX_TRUE@
 KDRIVEVESA_FALSE = @KDRIVEVESA_FALSE@
 KDRIVEVESA_TRUE = @KDRIVEVESA_TRUE@
 KDRIVE_CFLAGS = @KDRIVE_CFLAGS@
@@ -279,6 +283,8 @@
 MESA_SOURCE = @MESA_SOURCE@
 MFB_FALSE = @MFB_FALSE@
 MFB_TRUE = @MFB_TRUE@
+MISCUTILS_FALSE = @MISCUTILS_FALSE@
+MISCUTILS_TRUE = @MISCUTILS_TRUE@
 MISC_MAN_DIR = @MISC_MAN_DIR@
 MISC_MAN_SUFFIX = @MISC_MAN_SUFFIX@
 MITSHM_FALSE = @MITSHM_FALSE@
@@ -319,6 +325,7 @@
 RGB_DB = @RGB_DB@
 SCREENSAVER_FALSE = @SCREENSAVER_FALSE@
 SCREENSAVER_TRUE = @SCREENSAVER_TRUE@
+SERVERCONFIGdir = @SERVERCONFIGdir@
 SERVER_DEFINES = @SERVER_DEFINES@
 SET_MAKE = @SET_MAKE@
 SHELL = @SHELL@
@@ -565,6 +572,7 @@
                     api_loopback.c \
                     api_noop.c \
                     api_validate.c \
+					arrayobj.c \
                     attrib.c \
                     blend.c \
                     bufferobj.c \
@@ -603,6 +611,7 @@
                     points.c \
                     polygon.c \
                     rastpos.c \
+					rbadaptors.c \
                     renderbuffer.c \
                     state.c \
                     stencil.c \
@@ -676,6 +685,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/api_loopback.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/api_noop.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/api_validate.Plo@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/arrayobj.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/attrib.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/blend.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/bufferobj.Plo@am__quote@
@@ -714,6 +724,7 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/points.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/polygon.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/rastpos.Plo@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/rbadaptors.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/renderbuffer.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/state.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/stencil.Plo@am__quote@
