Index: xorg-server/hw/xfree86/common/xf86Cursor.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86Cursor.c	2006-09-18 17:27:03.000000000 -0400
+++ xorg-server/hw/xfree86/common/xf86Cursor.c	2006-10-13 16:03:16.000000000 -0400
@@ -214,6 +214,7 @@
   ScreenPtr   pCursorScreen;
   Bool        Switched;
   int         px, py;
+  int         sigstate;
 
   if (!pScr->vtSema || !mode || !pScr->SwitchMode)
     return FALSE;
@@ -233,6 +234,7 @@
   if (pScreen == pCursorScreen)
     miPointerPosition(&px, &py);
 
+  sigstate = xf86BlockSIGIO ();
   xf86EnterServerState(SETUP);
   Switched = (*pScr->SwitchMode)(pScr->scrnIndex, mode, 0);
   xf86EnterServerState(OPERATING);
@@ -270,6 +272,7 @@
       pScr->frameY1 = pScr->virtualY - 1;
     }
   }
+  xf86UnblockSIGIO (sigstate);
 
   if (pScr->AdjustFrame)
     (*pScr->AdjustFrame)(pScr->scrnIndex, pScr->frameX0, pScr->frameY0, 0);
