Index: xorg-server/hw/xfree86/os-support/bus/Pci.h
===================================================================
--- xorg-server.orig/hw/xfree86/os-support/bus/Pci.h	2006-10-14 21:57:12.000000000 -0700
+++ xorg-server/hw/xfree86/os-support/bus/Pci.h	2006-10-14 21:57:46.000000000 -0700
@@ -328,7 +328,7 @@
 #  define INCLUDE_XF86_MAP_PCI_MEM
 #  define INCLUDE_XF86_NO_DOMAIN
 # endif
-# if !defined(__FreeBSD__)
+# if !defined(__FreeBSD__) && !defined(linux)
 #  define ARCH_PCI_PCI_BRIDGE sparcPciPciBridge
 # endif
 #elif defined(__amd64__) || defined(__amd64)
