Index: hw/xfree86/os-support/linux/Makefile.am
===================================================================
--- hw/xfree86/os-support/linux/Makefile.am.orig	2007-02-04 07:41:11.000000000 +0100
+++ hw/xfree86/os-support/linux/Makefile.am	2007-02-04 08:12:21.000000000 +0100
@@ -6,7 +6,12 @@
 PLATFORM_INCLUDES = -I$(srcdir)/../shared
 endif
 if LINUX_ALPHA
-PLATFORM_PCI_SUPPORT = lnx_ev56.c \
+noinst_LTLIBRARIES += libev56.la
+libev56_la_CFLAGS = $(AM_CFLAGS) -mcpu=ev56
+libev56_la_SOURCES = lnx_ev56.c
+liblinux_la_LIBADD = libev56.la
+liblinux_la_DEPENDENCIES = libev56.la
+PLATFORM_PCI_SUPPORT = \
        $(srcdir)/lnx_axp.c \
        $(srcdir)/../shared/xf86Axp.c
 endif
