Index: hw/xfree86/os-support/linux/Makefile.in
===================================================================
--- hw/xfree86/os-support/linux/Makefile.in	(revision 3907)
+++ hw/xfree86/os-support/linux/Makefile.in	(working copy)
@@ -36,8 +36,9 @@
 POST_UNINSTALL = :
 build_triplet = @build@
 host_triplet = @host@
-@LNXACPI_TRUE@am__append_1 = -DHAVE_ACPI
-@LNXAPM_TRUE@am__append_2 = -DHAVE_APM
+@LINUX_ALPHA_TRUE@am__append_1 = libev56.la
+@LNXACPI_TRUE@am__append_2 = -DHAVE_ACPI
+@LNXAPM_TRUE@am__append_3 = -DHAVE_APM
 subdir = hw/xfree86/os-support/linux
 DIST_COMMON = $(srcdir)/Makefile.am $(srcdir)/Makefile.in
 ACLOCAL_M4 = $(top_srcdir)/aclocal.m4
@@ -56,7 +57,11 @@
 	$(top_builddir)/include/kdrive-config.h
 CONFIG_CLEAN_FILES =
 LTLIBRARIES = $(noinst_LTLIBRARIES)
-liblinux_la_LIBADD =
+libev56_la_LIBADD =
+am__libev56_la_SOURCES_DIST = lnx_ev56.c
+@LINUX_ALPHA_TRUE@am_libev56_la_OBJECTS = libev56_la-lnx_ev56.lo
+libev56_la_OBJECTS = $(am_libev56_la_OBJECTS)
+@LINUX_ALPHA_TRUE@am_libev56_la_rpath =
 am__liblinux_la_SOURCES_DIST = lnx_init.c lnx_video.c lnx_io.c \
 	lnx_kbd.c lnx_mouse.c lnx_pci.c lnx_agp.c lnx_kmod.c \
 	lnx_KbdMap.c $(srcdir)/../shared/bios_mmap.c \
@@ -65,12 +70,12 @@
 	$(srcdir)/../shared/sigio.c $(srcdir)/../shared/stdResource.c \
 	$(srcdir)/../shared/libc_wrapper.c \
 	$(srcdir)/../shared/at_scancode.c lnx_acpi.c lnx_apm.c \
-	lnx_ev56.c $(srcdir)/lnx_axp.c $(srcdir)/../shared/xf86Axp.c \
+	$(srcdir)/lnx_axp.c $(srcdir)/../shared/xf86Axp.c \
 	$(srcdir)/../shared/ia64Pci.c
 @LNXACPI_TRUE@am__objects_1 = lnx_acpi.lo lnx_apm.lo
 @LNXAPM_TRUE@am__objects_2 = lnx_apm.lo
 @LINUX_ALPHA_FALSE@@LINUX_IA64_TRUE@am__objects_3 = ia64Pci.lo
-@LINUX_ALPHA_TRUE@am__objects_3 = lnx_ev56.lo lnx_axp.lo xf86Axp.lo
+@LINUX_ALPHA_TRUE@am__objects_3 = lnx_axp.lo xf86Axp.lo
 am_liblinux_la_OBJECTS = lnx_init.lo lnx_video.lo lnx_io.lo lnx_kbd.lo \
 	lnx_mouse.lo lnx_pci.lo lnx_agp.lo lnx_kmod.lo lnx_KbdMap.lo \
 	bios_mmap.lo VTsw_usl.lo std_kbdEv.lo posix_tty.lo vidmem.lo \
@@ -88,8 +93,9 @@
 CCLD = $(CC)
 LINK = $(LIBTOOL) --tag=CC --mode=link $(CCLD) $(AM_CFLAGS) $(CFLAGS) \
 	$(AM_LDFLAGS) $(LDFLAGS) -o $@
-SOURCES = $(liblinux_la_SOURCES)
-DIST_SOURCES = $(am__liblinux_la_SOURCES_DIST)
+SOURCES = $(libev56_la_SOURCES) $(liblinux_la_SOURCES)
+DIST_SOURCES = $(am__libev56_la_SOURCES_DIST) \
+	$(am__liblinux_la_SOURCES_DIST)
 ETAGS = etags
 CTAGS = ctags
 DISTFILES = $(DIST_COMMON) $(DIST_SOURCES) $(TEXINFOS) $(EXTRA_DIST)
@@ -209,7 +215,6 @@
 FREETYPE_CFLAGS = @FREETYPE_CFLAGS@
 FREETYPE_LIBS = @FREETYPE_LIBS@
 FREETYPE_REQUIRES = @FREETYPE_REQUIRES@
-GLX_ARCH_DEFINES = @GLX_ARCH_DEFINES@
 GLX_DEFINES = @GLX_DEFINES@
 GLX_FALSE = @GLX_FALSE@
 GLX_TRUE = @GLX_TRUE@
@@ -236,8 +241,6 @@
 INT10_X86EMU_TRUE = @INT10_X86EMU_TRUE@
 KDRIVEFBDEV_FALSE = @KDRIVEFBDEV_FALSE@
 KDRIVEFBDEV_TRUE = @KDRIVEFBDEV_TRUE@
-KDRIVELINUX_FALSE = @KDRIVELINUX_FALSE@
-KDRIVELINUX_TRUE = @KDRIVELINUX_TRUE@
 KDRIVEVESA_FALSE = @KDRIVEVESA_FALSE@
 KDRIVEVESA_TRUE = @KDRIVEVESA_TRUE@
 KDRIVE_CFLAGS = @KDRIVE_CFLAGS@
@@ -290,8 +293,6 @@
 MESA_SOURCE = @MESA_SOURCE@
 MFB_FALSE = @MFB_FALSE@
 MFB_TRUE = @MFB_TRUE@
-MISCUTILS_FALSE = @MISCUTILS_FALSE@
-MISCUTILS_TRUE = @MISCUTILS_TRUE@
 MISC_MAN_DIR = @MISC_MAN_DIR@
 MISC_MAN_SUFFIX = @MISC_MAN_SUFFIX@
 MITSHM_FALSE = @MITSHM_FALSE@
@@ -332,7 +333,6 @@
 RGB_DB = @RGB_DB@
 SCREENSAVER_FALSE = @SCREENSAVER_FALSE@
 SCREENSAVER_TRUE = @SCREENSAVER_TRUE@
-SERVERCONFIGdir = @SERVERCONFIGdir@
 SERVER_DEFINES = @SERVER_DEFINES@
 SET_MAKE = @SET_MAKE@
 SHELL = @SHELL@
@@ -430,7 +430,7 @@
 XORG_BUS_SPARCPCI_TRUE = @XORG_BUS_SPARCPCI_TRUE@
 XORG_BUS_SPARC_FALSE = @XORG_BUS_SPARC_FALSE@
 XORG_BUS_SPARC_TRUE = @XORG_BUS_SPARC_TRUE@
-XORG_CFLAGS = @XORG_CFLAGS@ $(am__append_1) $(am__append_2)
+XORG_CFLAGS = @XORG_CFLAGS@ $(am__append_2) $(am__append_3)
 XORG_CORE_LIBS = @XORG_CORE_LIBS@
 XORG_FALSE = @XORG_FALSE@
 XORG_INCS = @XORG_INCS@
@@ -552,12 +552,16 @@
 target_alias = @target_alias@
 xglmoduledir = @xglmoduledir@
 xpconfigdir = @xpconfigdir@
-noinst_LTLIBRARIES = liblinux.la
-@LINUX_ALPHA_TRUE@PLATFORM_PCI_SUPPORT = lnx_ev56.c \
+noinst_LTLIBRARIES = liblinux.la $(am__append_1)
+@LINUX_ALPHA_TRUE@PLATFORM_PCI_SUPPORT = \
 @LINUX_ALPHA_TRUE@       $(srcdir)/lnx_axp.c \
 @LINUX_ALPHA_TRUE@       $(srcdir)/../shared/xf86Axp.c
 
 @LINUX_IA64_TRUE@PLATFORM_PCI_SUPPORT = $(srcdir)/../shared/ia64Pci.c
+@LINUX_ALPHA_TRUE@libev56_la_CFLAGS = $(AM_CFLAGS) -mcpu=ev56
+@LINUX_ALPHA_TRUE@libev56_la_SOURCES = lnx_ev56.c
+@LINUX_ALPHA_TRUE@liblinux_la_LIBADD = libev56.la
+@LINUX_ALPHA_TRUE@liblinux_la_DEPENDENCIES = libev56.la
 @LNXACPI_TRUE@ACPI_SOURCES = lnx_acpi.c lnx_apm.c
 @LNXAPM_TRUE@APM_SOURCES = lnx_apm.c
 liblinux_la_SOURCES = lnx_init.c lnx_video.c lnx_io.c lnx_kbd.c lnx_mouse.c \
@@ -634,6 +638,8 @@
 	  echo "rm -f \"$${dir}/so_locations\""; \
 	  rm -f "$${dir}/so_locations"; \
 	done
+libev56.la: $(libev56_la_OBJECTS) $(libev56_la_DEPENDENCIES) 
+	$(LINK) $(am_libev56_la_rpath) $(libev56_la_LDFLAGS) $(libev56_la_OBJECTS) $(libev56_la_LIBADD) $(LIBS)
 liblinux.la: $(liblinux_la_OBJECTS) $(liblinux_la_DEPENDENCIES) 
 	$(LINK)  $(liblinux_la_LDFLAGS) $(liblinux_la_OBJECTS) $(liblinux_la_LIBADD) $(LIBS)
 
@@ -648,12 +654,12 @@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/bios_mmap.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/ia64Pci.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/libc_wrapper.Plo@am__quote@
+@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/libev56_la-lnx_ev56.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/lnx_KbdMap.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/lnx_acpi.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/lnx_agp.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/lnx_apm.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/lnx_axp.Plo@am__quote@
-@AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/lnx_ev56.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/lnx_init.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/lnx_io.Plo@am__quote@
 @AMDEP_TRUE@@am__include@ @am__quote@./$(DEPDIR)/lnx_kbd.Plo@am__quote@
@@ -689,6 +695,13 @@
 @AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
 @am__fastdepCC_FALSE@	$(LTCOMPILE) -c -o $@ $<
 
+libev56_la-lnx_ev56.lo: lnx_ev56.c
+@am__fastdepCC_TRUE@	if $(LIBTOOL) --tag=CC --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(libev56_la_CFLAGS) $(CFLAGS) -MT libev56_la-lnx_ev56.lo -MD -MP -MF "$(DEPDIR)/libev56_la-lnx_ev56.Tpo" -c -o libev56_la-lnx_ev56.lo `test -f 'lnx_ev56.c' || echo '$(srcdir)/'`lnx_ev56.c; \
+@am__fastdepCC_TRUE@	then mv -f "$(DEPDIR)/libev56_la-lnx_ev56.Tpo" "$(DEPDIR)/libev56_la-lnx_ev56.Plo"; else rm -f "$(DEPDIR)/libev56_la-lnx_ev56.Tpo"; exit 1; fi
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	source='lnx_ev56.c' object='libev56_la-lnx_ev56.lo' libtool=yes @AMDEPBACKSLASH@
+@AMDEP_TRUE@@am__fastdepCC_FALSE@	DEPDIR=$(DEPDIR) $(CCDEPMODE) $(depcomp) @AMDEPBACKSLASH@
+@am__fastdepCC_FALSE@	$(LIBTOOL) --tag=CC --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(libev56_la_CFLAGS) $(CFLAGS) -c -o libev56_la-lnx_ev56.lo `test -f 'lnx_ev56.c' || echo '$(srcdir)/'`lnx_ev56.c
+
 bios_mmap.lo: $(srcdir)/../shared/bios_mmap.c
 @am__fastdepCC_TRUE@	if $(LIBTOOL) --tag=CC --mode=compile $(CC) $(DEFS) $(DEFAULT_INCLUDES) $(INCLUDES) $(AM_CPPFLAGS) $(CPPFLAGS) $(AM_CFLAGS) $(CFLAGS) -MT bios_mmap.lo -MD -MP -MF "$(DEPDIR)/bios_mmap.Tpo" -c -o bios_mmap.lo `test -f '$(srcdir)/../shared/bios_mmap.c' || echo '$(srcdir)/'`$(srcdir)/../shared/bios_mmap.c; \
 @am__fastdepCC_TRUE@	then mv -f "$(DEPDIR)/bios_mmap.Tpo" "$(DEPDIR)/bios_mmap.Plo"; else rm -f "$(DEPDIR)/bios_mmap.Tpo"; exit 1; fi
Index: hw/xfree86/os-support/linux/Makefile.am
===================================================================
--- hw/xfree86/os-support/linux/Makefile.am	(revision 3907)
+++ hw/xfree86/os-support/linux/Makefile.am	(working copy)
@@ -4,7 +4,12 @@
 PLATFORM_PCI_SUPPORT = $(srcdir)/../shared/ia64Pci.c
 endif
 if LINUX_ALPHA
-PLATFORM_PCI_SUPPORT = lnx_ev56.c \
+noinst_LTLIBRARIES += libev56.la
+libev56_la_CFLAGS = $(AM_CFLAGS) -mcpu=ev56
+libev56_la_SOURCES = lnx_ev56.c
+liblinux_la_LIBADD = libev56.la
+liblinux_la_DEPENDENCIES = libev56.la
+PLATFORM_PCI_SUPPORT = \
        $(srcdir)/lnx_axp.c \
        $(srcdir)/../shared/xf86Axp.c
 endif
