Index: xorg-server/hw/xfree86/int10/Makefile.am
===================================================================
--- xorg-server.orig/hw/xfree86/int10/Makefile.am	2007-02-06 00:16:58.000000000 +0100
+++ xorg-server/hw/xfree86/int10/Makefile.am	2007-02-06 00:22:47.000000000 +0100
@@ -7,16 +7,42 @@
 libint10_la_LDFLAGS = -avoid-version
 
 if INT10_VM86
-AM_CFLAGS = -D_PC -D_VM86_LINUX $(XORG_CFLAGS) $(EXTRA_CFLAGS)
+module_LTLIBRARIES += libvm86.la libx86emu.la
+AM_CFLAGS = -D_PC $(XORG_CFLAGS) $(EXTRA_CFLAGS)
 INCLUDES = $(XORG_INCS)
+libint10_la_CPPFLAGS = $(AM_CPPFLAGS) -DHAVE_SYSV_IPC -DDoSubModules
 libint10_la_SOURCES = \
-	helper_exec.c \
-	helper_mem.c \
 	pci.c \
-	xf86int10.c \
 	xf86int10module.c \
-	$(srcdir)/../os-support/linux/int10/vm86/linux_vm86.c \
+	helper_mem.c \
 	$(srcdir)/../os-support/linux/int10/linux.c
+libvm86_la_LDFLAGS = -avoid-version
+libvm86_la_CPPFLAGS = \
+	$(AM_CPPFLAGS) \
+	-DMOD_NAME=vm86 \
+	-DHAVE_SYSV_IPC \
+	-DSHOW_ALL_DEVICES \
+	-D_VM86_LINUX
+libvm86_la_SOURCES = \
+	xf86int10module.c \
+	$(srcdir)/../os-support/linux/int10/vm86/linux_vm86.c \
+	helper_exec.c \
+	xf86int10.c
+libx86emu_la_LDFLAGS = -avoid-version
+libx86emu_la_CPPFLAGS = \
+	$(AM_CPPFLAGS) \
+	-DMOD_NAME=x86emu \
+	-DHAVE_SYSV_IPC \
+	-DSHOW_ALL_DEVICES \
+	-D_X86EMU \
+	-DNO_SYS_HEADERS \
+	-I$(srcdir)/../x86emu
+libx86emu_la_SOURCES = \
+	xf86int10module.c \
+	xf86x86emu.c \
+	helper_exec.c \
+	xf86int10.c \
+	x86emu.c
 endif
 
 if INT10_X86EMU
Index: xorg-server/hw/xfree86/int10/xf86int10.c
===================================================================
--- xorg-server.orig/hw/xfree86/int10/xf86int10.c	2007-02-06 00:23:05.000000000 +0100
+++ xorg-server/hw/xfree86/int10/xf86int10.c	2007-02-06 00:23:35.000000000 +0100
@@ -17,7 +17,7 @@
 
 #define REG pInt
 
-xf86Int10InfoPtr Int10Current = NULL;
+// xf86Int10InfoPtr Int10Current = NULL;
 
 static int int1A_handler(xf86Int10InfoPtr pInt);
 #ifndef _PC
Index: xorg-server/hw/xfree86/int10/helper_mem.c
===================================================================
--- xorg-server.orig/hw/xfree86/int10/helper_mem.c	2007-02-06 00:23:48.000000000 +0100
+++ xorg-server/hw/xfree86/int10/helper_mem.c	2007-02-06 00:24:15.000000000 +0100
@@ -23,6 +23,8 @@
 
 #define REG pInt
 
+xf86Int10InfoPtr Int10Current = NULL;
+
 typedef enum {
     OPT_NOINT10,
     OPT_INIT_PRIMARY,
