Index: xorg-server/hw/xfree86/int10/Makefile.am
===================================================================
--- xorg-server.orig/hw/xfree86/int10/Makefile.am	2007-02-10 17:52:53.000000000 +0100
+++ xorg-server/hw/xfree86/int10/Makefile.am	2007-02-10 17:55:26.000000000 +0100
@@ -18,12 +18,42 @@
 endif
 
 if INT10_VM86
-AM_CFLAGS = $(I386_VIDEO_CFLAGS) -D_VM86_LINUX $(XORG_CFLAGS) $(EXTRA_CFLAGS)
+module_LTLIBRARIES += libvm86.la libx86emu.la
+AM_CFLAGS = $(I386_VIDEO_CFLAGS) $(XORG_CFLAGS) $(EXTRA_CFLAGS)
 INCLUDES = $(XORG_INCS)
+libint10_la_CPPFLAGS = $(AM_CPPFLAGS) -DHAVE_SYSV_IPC -DDoSubModules
 libint10_la_SOURCES = \
-	$(COMMON_SOURCES) \
-	$(srcdir)/../os-support/linux/int10/vm86/linux_vm86.c \
+	pci.c \
+	xf86int10module.c \
+	helper_mem.c \
 	$(srcdir)/../os-support/linux/int10/linux.c
+libvm86_la_LDFLAGS = -avoid-version
+libvm86_la_CPPFLAGS = \
+	$(AM_CPPFLAGS) \
+	-DMOD_NAME=vm86 \
+	-DHAVE_SYSV_IPC \
+	-DSHOW_ALL_DEVICES \
+	-D_VM86_LINUX
+libvm86_la_SOURCES = \
+	xf86int10module.c \
+	$(srcdir)/../os-support/linux/int10/vm86/linux_vm86.c \
+	helper_exec.c \
+	xf86int10.c
+libx86emu_la_LDFLAGS = -avoid-version
+libx86emu_la_CPPFLAGS = \
+	$(AM_CPPFLAGS) \
+	-DMOD_NAME=x86emu \
+	-DHAVE_SYSV_IPC \
+	-DSHOW_ALL_DEVICES \
+	-D_X86EMU \
+	-DNO_SYS_HEADERS \
+	-I$(srcdir)/../x86emu
+libx86emu_la_SOURCES = \
+	xf86int10module.c \
+	xf86x86emu.c \
+	helper_exec.c \
+	xf86int10.c \
+	x86emu.c
 endif
 
 if INT10_X86EMU
Index: xorg-server/hw/xfree86/int10/xf86int10.c
===================================================================
--- xorg-server.orig/hw/xfree86/int10/xf86int10.c	2007-02-10 17:52:53.000000000 +0100
+++ xorg-server/hw/xfree86/int10/xf86int10.c	2007-02-10 17:54:11.000000000 +0100
@@ -16,7 +16,7 @@
 
 #define REG pInt
 
-xf86Int10InfoPtr Int10Current = NULL;
+// xf86Int10InfoPtr Int10Current = NULL;
 
 static int int1A_handler(xf86Int10InfoPtr pInt);
 #ifndef _PC
Index: xorg-server/hw/xfree86/int10/helper_mem.c
===================================================================
--- xorg-server.orig/hw/xfree86/int10/helper_mem.c	2007-02-10 17:52:53.000000000 +0100
+++ xorg-server/hw/xfree86/int10/helper_mem.c	2007-02-10 17:54:31.000000000 +0100
@@ -22,6 +22,8 @@
 
 #define REG pInt
 
+xf86Int10InfoPtr Int10Current = NULL;
+
 typedef enum {
     OPT_NOINT10,
     OPT_INIT_PRIMARY,
@@ -429,3 +431,9 @@
     pInt->BIOSseg = cs;
     return TRUE;
 }
+
+xf86Int10InfoPtr
+xf86InitInt10(int entityIndex)
+{
+    return xf86ExtendedInitInt10(entityIndex, 0);
+}
Index: xorg-server/hw/xfree86/int10/helper_exec.c
===================================================================
--- xorg-server.orig/hw/xfree86/int10/helper_exec.c	2007-02-10 17:54:35.000000000 +0100
+++ xorg-server/hw/xfree86/int10/helper_exec.c	2007-02-10 17:54:50.000000000 +0100
@@ -684,8 +684,3 @@
 }
 #endif
 
-xf86Int10InfoPtr
-xf86InitInt10(int entityIndex)
-{
-    return xf86ExtendedInitInt10(entityIndex, 0);
-}
