--- a/configure.ac
+++ b/configure.ac
@@ -849,7 +849,7 @@ if test "x$WITH_SYSTEMD_DAEMON" = xyes;
 		AC_MSG_ERROR([systemd support requested but no library has been found])
 	fi
 	AC_DEFINE(HAVE_SYSTEMD_DAEMON, 1, [Define to 1 if libsystemd-daemon is available])
-	REQUIRED_LIBS="$REQUIRED_LIBS libsystemd-daemon"
+	REQUIRED_LIBS="$REQUIRED_LIBS libsystemd"
 fi
 AM_CONDITIONAL([HAVE_SYSTEMD_DAEMON], [test "x$HAVE_SYSTEMD_DAEMON" = "xyes"])
 
