From f8079585cd68ad2e742d611bdf9908dc6db5a4a9 Mon Sep 17 00:00:00 2001
From: Fedora X Ninjas <x@fedoraproject.org>
Date: Fri, 10 Jul 2009 16:26:25 +0200
Subject: [PATCH] Add vboxvideo to the autodetection routine

---
 hw/xfree86/common/xf86AutoConfig.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

Index: xorg-server/hw/xfree86/common/xf86AutoConfig.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86AutoConfig.c
+++ xorg-server/hw/xfree86/common/xf86AutoConfig.c
@@ -207,6 +207,7 @@
 	case 0x100c:		    driverList[0] = "tseng";	break;
 	case 0x1106:		    driverList[0] = "openchrome"; break;
 	case 0x15ad:		    driverList[0] = "vmware";	break;
+	case 0x80ee:		    driverList[0] = "vboxvideo"; break;
 	case 0x18ca:
 	    if (dev->device_id == 0x47)
 		driverList[0] = "xgixp";
