From a4c328fe2f65d1bacd5593233277b0ea4eb9af69 Mon Sep 17 00:00:00 2001
From: Julien Cristau <jcristau@debian.org>
Date: Sun, 13 Sep 2009 11:52:37 +0200
Subject: [PATCH] render: return the supported version rather than just passing the proto's version

---
 render/render.c |    7 +++++--
 1 files changed, 5 insertions(+), 2 deletions(-)

Index: xorg-server/render/render.c
===================================================================
--- xorg-server.orig/render/render.c
+++ xorg-server/render/render.c
@@ -54,6 +54,9 @@
 #define UINT32_MAX 0xffffffffU
 #endif
 
+#define SERVER_RENDER_MAJOR 0
+#define SERVER_RENDER_MINOR 10
+
 static int ProcRenderQueryVersion (ClientPtr pClient);
 static int ProcRenderQueryPictFormats (ClientPtr pClient);
 static int ProcRenderQueryPictIndexValues (ClientPtr pClient);
@@ -271,8 +274,8 @@
     rep.type = X_Reply;
     rep.length = 0;
     rep.sequenceNumber = client->sequence;
-    rep.majorVersion = RENDER_MAJOR;
-    rep.minorVersion = RENDER_MINOR;
+    rep.majorVersion = SERVER_RENDER_MAJOR;
+    rep.minorVersion = SERVER_RENDER_MINOR;
     if (client->swapped) {
     	swaps(&rep.sequenceNumber, n);
     	swapl(&rep.length, n);
