From cbaf21a87c207300ab5a1e7de234e03baa0edf80 Mon Sep 17 00:00:00 2001
From: Julien Cristau <jcristau@debian.org>
Date: Sun, 8 Mar 2009 22:19:10 +0100
Subject: [PATCH] Turn on ModeDebug by default

---
 hw/xfree86/modes/xf86Crtc.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

Index: xorg-server/hw/xfree86/modes/xf86Crtc.c
===================================================================
--- xorg-server.orig/hw/xfree86/modes/xf86Crtc.c
+++ xorg-server/hw/xfree86/modes/xf86Crtc.c
@@ -2356,7 +2356,7 @@
 			scrn->options,
 			config->options);
     config->debug_modes = xf86ReturnOptValBool (config->options,
-						OPTION_MODEDEBUG, FALSE);
+						OPTION_MODEDEBUG, TRUE);
 
     if (scrn->display->virtualX)
 	width = scrn->display->virtualX;
